### ISSUE TITLE
Union/overlay "filesystem" using LD_PRELOAD

### ISSUE THREAD

**Issue by probonopd** [probonopd/AppImageKit#267]
We experimented with this for klik 9 years ago:
https://www.winehq.org/pipermail/wine-devel/2007-November/060627.html

Someone else seems to have had some success with this idea:
http://algoholic.eu/unionfs_by_intercept/

**Comment by stsp** [comment]
> I don't know, but I'd be interested in the results. Lets continue this discussion in
> probonopd/AppImageKit#267

Just a quick look:

unionfs_by_intercept seems to be only intercepting
syscalls but not library calls. The problem with this is
that glibc calls the syscalls internally, you can't intercept
those. So when you call fopen(), you can't count on
intercepting open() - it won't work. You need to intercept
all library calls too.

preload.c is definitely interesting, but, being just a
part of other project, I wonder how generic/configurable
is it. I haven't tried it.

https://wiki.debian.org/FakeRoot is also an interesting
project in that regard, but it also is targeted for the
very specific use case.

unionfs-fuse is absolutely great, but IIRC it starts a
server process per every mount point, and slows down
things considerably (I mean, really-really considerably).

There was also a ptrace-based FS redirector somewhere.
Have to google for it a bit more.

user-union (which I co-authored) was the most
feature-rich and fast at time, but now I checked and
it doesn't even build easily. :( I'll try to find some time
to get it fixed.

**Comment by stsp** [comment]
So when I was dealing with this (and that was long ago)
I was combining unionfs-fuse for non performance-critical
parts, with user-union for performance-critical parts.
But many other projects seems to have emerged since that...

**Comment by stsp** [comment]
> http://www.dwheeler.com/user-union/ (thanks for the hint @stsp)

The interesting thing is that you already had it in your
repository even before my hint.
I don't know if you succeeded with building it (the problem
seems to depend on glibc version), but if you didn't, you
need to do:
./configure --disable-private-libc-namespace

Then it should build and work.

**Comment by probonopd** [comment]
Thanks. can you give your opinion on [preload.c](https://github.com/mikix/deb2snap/blob/master/src/preload.c) vs. user-union, do you see clear upsides/downsides of one over the other?

**Comment by stsp** [comment]
Unfortunately I don't see a list of features of
preload.c (like user-union's man page that I suggest
you to read for a basic idea about the feature set), and
I have not tried it myself, so I can only very basically
judge on the code look.
I wonder if preload.c is about FS overlaying. For example
I can see the FS redirection code there, but I can't see
the FS overlay code. So if I am not mistaken, the goal of
preload.c is just a very small subset of what user-union can
do. With preload.c you can create the private sandbox,
which is great. You can do the same with fakechroot too.
But with user-union you can do things like unionfs-fuse does:
mount multiple directories into a single mount-point, one on
top of another in a stack. If you don't need that functionality,
then user-union and unionfs-fuse are not your friends, but
if you need that, then it is exactly their task.

**Comment by stsp** [comment]
So if we are talking only about the FS redirection
and not union mounts, then it is entirely possible that
preload.c does things well. It looks fairly complete in
that regard, lots of syscalls and library calls wrapped,
maybe as much as user-union does, maybe even more.
It would be hard name the winner here unless you
need something what user-union is really designed for.

**Comment by stsp** [comment]
Well, as long as thsi ticket is called "Union/overlay "filesystem" using LD_PRELOAD"
and you already experimented with preload.c, my question is:
does it really do overlays for you? I don't see such code in there!

**Comment by probonopd** [comment]
Kinda: I can use it to "overlay" the contents of the AppImage over / with the result that if I launch $APPDIR/usr/bin/foo and it calls /usr/bin/bar, then bar works. But strangely it seems to be half-implemented because a directory listing or file open dialog box for /usr/bin will ONLY show the contents of $APPDIR/usr/bin but NOT in combination with the real /usr/bin... bug or feature or just not implemented?

**Comment by stsp** [comment]
> Kinda: I can use it to "overlay" the contents of the AppImage over / with the
> result that if I launch $APPDIR/usr/bin/foo and it calls /usr/bin/bar, then bar works.

What is "then bar works"?
Does $APPDIR/usr/bin/bar work or /usr/bin/bar works?
A big difference.

**Comment by probonopd** [comment]
/usr/bin/bar works; clarified my comment above.

**Comment by stsp** [comment]
> /usr/bin/bar works; clarified my comment above.

Is this really what you want?
With user-union you can get:
1. Attempt to run $APPDIR/usr/bin/bar
2. If 1 does not exist, the lower layer is tried, so /usr/bin/bar is started.

If it doesn't run $APPDIR/usr/bin/bar then there is
no overlay at all (as the code suggests anyway)

**Comment by probonopd** [comment]
I think I like user-union better, but it is much more complicated to use. What is great about preload.c is that you export one LD_PRELOAD variable and one UNION_PRELOAD variable and be done with it. In contrast, user-union comes with a convoluted bash wrapper script.

**Comment by stsp** [comment]
So what kind of overlay is this?
Am I right that you mount
$APPDIR to /
and if progs execute /usr/bin/bar, the look-up
in $APPDIR overlay is not done, and /usr/bin/bar
is executed as is?

**Comment by probonopd** [comment]
I would like $APPDIR to be overlayed to /
* If both /usr/bin/foo and $APPDIR/usr/bin/foo exist and /usr/bin/foo is accessed, then it should be used from $APPDIR/usr/bin/foo
* If only $APPDIR/usr/bin/foo exists and  /usr/bin/foo is accessed, then it should be used from /usr/bin/foo
* If only /usr/bin/foo exists and  /usr/bin/foo is accessed, then it should be used from /usr/bin/foo

**Comment by stsp** [comment]
> I would like $APPDIR to be overlayed to /

user-union supports mount points. I.e. you can overlay
$APPDIR with / and put the resulting mount to /mnt/overlay.

> If only $APPDIR/usr/bin/foo exists and /usr/bin/foo is accessed,
> then it should be used from /usr/bin/foo

Are you sure?
I think $APPDIR/usr/bin/foo should be used in this case.

> What is great about preload.c is that you export one LD_PRELOAD
> variable and one UNION_PRELOAD variable and be done with it.

That's because user-union supports a very complex
mounts. For example, you can have any mount of
R/O underlays with 1 or 0 overlay over them.
Also you can have multiple mounts at once.
Such flexibility requires setting many env vars
unfortunately, so it was simpler to add the script.

> half-implemented because a directory listing or file open dialog box for /usr/bin
> will ONLY show the contents of $APPDIR/usr/bin but NOT in combination with
> the real /usr/bin... bug or feature or just not implemented?

Seems not implemented, same in user-union. :)
File listing is a quite difficult to implement feature.

**Comment by probonopd** [comment]
> Are you sure?
> I think $APPDIR/usr/bin/foo should be used in this case.

Yes, I'm sure user-union is much more powerful, maybe too powerful for our simple usecase?
But maybe we could do a simplified subset of the functionality, with virtually no runtime options?
Actually that would be best, just set LD_PRELOAD and the rest is all hardcoded in the lib.

**Comment by stsp** [comment]
So I guess if you have just one mount
with one overlay and one underlay, and
the mount-point is equal to the underlay path,
then user-union is very unlikely to offer anything
over preload.c.
This is what you can do with
user-union -a $APPDIR /

To avoid the script, you can do the trick:
user-union -a $APPDIR / -n

-n will just print the needed env vars and exit.
You can set those vars yourself and use LD_PRELOAD then.

**Comment by probonopd** [comment]
What do you think about https://github.com/proot-me/PRoot?
http://www.slideshare.net/cvinc02/proot-improvedkernelcompat

**Comment by stsp** [comment]
If you have just the basic requirements and
preload.c is enough, why to search for more?

> What do you think about https://github.com/proot-me/PRoot?

The description says:
chroot, mount --bind, and binfmt_misc without privilege/setup

Doesn't sound like it supports overlay mounts.
I haven't really looked into it.

**Comment by probonopd** [comment]
> If you have just the basic requirements and preload.c is enough, why to search for more?

I just like to understand the options we have :-)

**Comment by stsp** [comment]
Then you can try the -n trick of user-union
to get the needed vars printed, recreate them
and compare to preload.c.

**Comment by probonopd** [comment]
Found an issue with `preload.c`:

This recipe __does not work correctly__ with `union: true` but __does__ with `binpatch: true`:

```
app: OpenSCAD
binpatch: true

ingredients:
  packages:
    - openscad-nightly
  dist: trusty
  sources: 
    - deb http://archive.ubuntu.com/ubuntu/ trusty main
    - deb http://download.opensuse.org/repositories/home:/t-paul/xUbuntu_14.04/ ./

script:
  - ls
```

The difference is that with `binpatch: true` the examples are shown in this window correctly:

![screenshot_2016-11-13_19-17-35](https://cloud.githubusercontent.com/assets/2480569/20247786/fa654c1e-a9d5-11e6-8011-a7d2146e3acc.png)

With `union: true` the examples are not shown and we get `Error reading examples.json: examples.json: cannot open file`.

**Comment by TheAssassin** [comment]
We should try to find out why. Binpatch is pretty much a hack, using the union file system is cleaner and we don't have to apply this `cd .../usr` hack.

**Comment by probonopd** [comment]
Also check https://github.com/Barthalion/hai which uses union-fuse and fakechroot.
